### PR TITLE
feat: add softmax examples

### PR DESCRIPTION
**Pull request by jlonge4**
### Description of changes:
Adding pytorch / NKI examples of implementing softmax

### Testing:

Please see detailed unit test requirements in the [CONTRIBUTING.md](https://github.com/aws-neuron/nki-samples/blob/main/CONTRIBUTING.md)

- [ ] The change is covered by numeric check using `nki.baremetal`
- [ ] The change is covered by performance benchmark test using `nki.benchmark`
- [ ] The change is covered by end-to-end integration test

### Pull Request Checklist

- [ ] I have filled in all the required field in the template
- [ ] I have tested locally that all the tests pass
- [ ] By submitting this pull request, I confirm that my contribution is made under the terms of the MIT-0 license.

### CODE REVIEW

File: src/nki_samples/tutorials/softmax/softmax_nki_kernels.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  # Since we're not reducing across the first dimension
+  # Tiles can be processed independently
+  for i in nl.affine_range(math.ceil(a_tensor.shape[0]/128)):
+
+    # Load input data from external memory to on-chip memory
+    a_tile = nl.load(a_tensor[i * 128 + ix, iy],
+                    mask=(i * 128 + ix < num_rows))
+
+    # Find max and subtract from each value to ensure numerical stability
+    max_vals = nl.max(a_tile, axis=[1], keepdims=True)
```
Comment (JonathanHenson) [inline src/nki_samples/tutorials/softmax/softmax_nki_kernels.py:29]: mask needs to be propagated to each of these operations:

nl.max(a_tile, axis=[1], keepdims=True, mask=...)
nl.subtract(a_tile, max_vales, mask=...)

**Review by JonathanHenson — CHANGES_REQUESTED** [review]

File: src/nki_samples/tutorials/softmax/softmax_nki_kernels.py (hunk 1 above)
Comment (jlonge4) [reply src/nki_samples/tutorials/softmax/softmax_nki_kernels.py]: Done and re-tested

Comment (JonathanHenson) [reply src/nki_samples/tutorials/softmax/softmax_nki_kernels.py]: I'm not still not seeing mask propogated in nl.max, nl.subtract, nl.exp etc?

**Review by aws-qieqingy — APPROVED** [review]

**Review by JonathanHenson — APPROVED** [review]
